src/helpers/reciever.js

```diff
@@ -0,0 +1,330 @@
+class Receiver {
+  constructor(DataBaseName) {
+    this.DataBaseName = DataBaseName;
+    this.CWAData = null;
+    this.ReleaseMap = {};//hash map
+    this.latestItem = null;
+    var date = new Date();
+    this.year = date.getFullYear().toString();
+    let cacheLatestItem = this.cacheGetLatestItem();
+    if (cacheLatestItem !== undefined) {
+      this.CWAData = this.cacheGetAllData();
+      this.latestItem = cacheLatestItem;
+    }
+    var tempReleaseList = this.cacheGetReleaseList()
+    if(tempReleaseList !==null){
+      this.ReleaseMap = tempReleaseList
+    }
+  }
+
+  /*update()
+  Desc: This function async. updates the local storage by comparing latest 
+  item in dynamo and the latest item in local storage. If found a difference
+  it updates them by calling getBatchItem to retrieve new data from dynamo
+  Return: if the sync is successful or not
+  */
+  async update() {
+    // check the latest hash from cache
+    try {
+      let dynamoLatestItem = await this.getLatestItem();
+      let DynamoHash = dynamoLatestItem[HASH];
+      // ask dynamo what is the lastest hash it received
+      let cacheLatestItem = this.cacheGetLatestItem(); //[HASH].S //rename to lastest hash
+      let cacheLatestHash = "";
+      if (cacheLatestItem === null) {
+        // no cache found, pull every thing and set
+        this.CWAData = await this.getAllItems();
+        this.latestItem = dynamoLatestItem;
+        this.cacheSaveData();
+        return true;
+      } else {
+        cacheLatestHash = cacheLatestItem[HASH];
+        this.CWAData = this.cacheGetAllData();
+      }
+
+      if (DynamoHash === cacheLatestHash) {
+        return this.updateReleases() // already synced
+      } else if (
+        parseInt(dynamoLatestItem[COMMIT_DATE]) >=
+        parseInt(cacheLatestItem[COMMIT_DATE])
+      ) {
+        /// if hashes dont match call getBatchItem with local hash and update local hash with new data
+        var newItems = await this.getBatchItem(
+          cacheLatestItem[COMMIT_DATE],
+          dynamoLatestItem[COMMIT_DATE]
+        );
+        Object.keys(newItems).forEach((key) => {
+          if (this.CWAData[key] === undefined) {
+            // new testCase
+            this.CWAData[key] = {};
+          }
+          //already have this testCase
+          Object.keys(this.CWAData[key]).forEach((metric) => {
+            if (this.CWAData[key][metric] === undefined) {
+              // new metric
+              this.CWAData[key][metric] = [];
+            }
+            //already have this metric
+            this.CWAData[key][metric].push(...newItems[key][metric]);
+          });
+        });
+        this.latestItem = dynamoLatestItem;
+        this.cacheSaveData();
+        return this.updateReleases() // now synced
+      }
+      // website is ahead of dynamo
+      //clear cache and retry 
+      this.cacheClear()
+      this.update()
+    } catch (err) {
+      console.log(`ERROR:${err}`);
+      alert(`ERROR:${err}`);
```

Comment:
Do we really want to alert to the browser directly with the error message?

Reply:
Good point, should I still console.log it? I kinda want the user to have some way of adding the error code to a ticket if there is a problem.

Reply:
I think we should still at least log the error, there should be some output that tells them what went wrong.